### ISSUE TITLE
Custom components with ComponentState not compatible with rx.foreach

### ISSUE THREAD

**Issue by MarcoTuc** [reflex-dev/reflex#3718]
When a Custom component is created with a custom state, if declared via rx.foreach, then all the created components will share the same state. Below an example that works as I explained on reflex 0.5.7
```
class ReusableCounter(rx.ComponentState):
    count: int = 0

    def increment(self):
        self.count += 1

    def decrement(self):
        self.count -= 1

    @classmethod
    def get_component(cls, **props):
        return rx.hstack(
            rx.button("Decrement", on_click=cls.decrement),
            rx.text(cls.count),
            rx.button("Increment", on_click=cls.increment),
            **props,
        )


reusable_counter = ReusableCounter.create


def multiple_counters():
    return rx.foreach(
        rx.Var.range(3),
        lambda i: rx.hstack(
            rx.text(f"Counter {i}"),
            reusable_counter(),
        )
    )
```

**Comment by benedikt-bartscher** [comment]
Dynamic States are currently not possible in reflex because all state names need to be known at compile time. I would love to see a solution for dynamic states in general and later use it in ComponentState.

I did some exploration in #3671 and #3683

**Comment by yutayouguan** [comment]
has any  progress on the issue , it puzzled for a long time,

**Comment by paoloemilioserra** [comment]
looking for an update on this eventually

**Comment by jq6l43d1** [comment]
This is a bug I would also love to see fixed. Using rx.foreach with rx.ComponentState would be very useful.

I wrote a website that shows info cards for all of my VMs. VMs can be created or deleted at any time. I wanted to each VM to have it's own VM card component with it's own component state. I wrote way too much code before I realized that this isn't possible.

Adding a note about this limitation to the documentation page [here](https://reflex.dev/docs/substates/component-state/) would prevent the unpleasant surprise when the following error pops up unexpectedly:

TypeError: Using a ComponentState as `render_fn` inside `rx.foreach` is not supported yet.

**Comment by benedikt-bartscher** [comment]
@jq6l43d1 it is not really a bug, it is a missing feature. Currently every `rx.State` subclass will be instantiated max once per session. This feature would require some heavy refactoring which @abulvenz and I tried some time ago (https://github.com/reflex-dev/reflex/pull/3671 and https://github.com/reflex-dev/reflex/pull/3683)

**Comment by jq6l43d1** [comment]
@benedikt-bartscher Thanks for looking into this. As a workaround I just created a bunch of component states. I usually have around 10 VMs so I made like 20 states like so:

```
    vm_00 = reusable_vm_card(vm_index=0)
    vm_01 = reusable_vm_card(vm_index=1)
    vm_02 = reusable_vm_card(vm_index=2)
```

Then in the component state I set the name of the VM to the value of my list_of_vms variable that has the index of vm_index or I simply return if it would be out of index. Then I use rx.cond to render the VM info card if the value of vm_name isn't blank or render rx.fragment if it is blank. The only down side is I have to make sure I'm creating more component states than I will need. And it's more work than if we could use rx.foreach.

**Comment by johschmidt42** [comment]
Is there a current workaround?

Like @jq6l43d1 I built way too much before realising that I can't build my data table the way I wanted. With a dynamic state variable (e.g. list of strings), one must use `rx.foreach` which currently doesn't work with components based on `rx.Component`. But currently I don't see a way around it?

For example: I fetch data from an API and store that in a global state class (assuming a list of strings with variable length). For each item, I want a custom component (e.g. copy to clipboard button). What would be your suggestion? Thanks in advance!

**Comment by benedikt-bartscher** [comment]
> Is there a current workaround?

No, at least as far as I know. ComponentState is compile-time, reflex can't handle dynamic states at runtime, atm.

They only "workaround" I can imagine is to move all Vars (base+computed) from your ComponentState to a dataclass or rx.Base and use it in a list/dict state var. However, this has bad ergonomics and performance.
